File: cpp/examples/handoff.cpp
// C++ equivalent of examples/test-handoff-hmac.js and examples/test-handoff-ed25519.js.
// Reads the same repo-root .env as the Node examples (walks up from the working directory).
//
//   ./build/handoff_example hmac
//   ./build/handoff_example ed25519

#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>

#include "danumai/partner_signed_client.hpp"

namespace fs = std::filesystem;
using namespace danumai::partner;

namespace {

std::string trim(const std::string& s) {
    const auto start = s.find_first_not_of(" \t\r\n");
    if (start == std::string::npos) return {};
    const auto end = s.find_last_not_of(" \t\r\n");
    return s.substr(start, end - start + 1);
}

/// Minimal .env loader — walks up from the working directory to the first .env; real env vars win.
void load_env() {
    for (fs::path dir = fs::current_path();; dir = dir.parent_path()) {
        const fs::path file = dir / ".env";
        if (fs::is_regular_file(file)) {
            std::ifstream in(file);
            for (std::string line; std::getline(in, line);) {
                line = trim(line);
                const auto eq = line.find('=');
                if (line.empty() || line[0] == '#' || eq == std::string::npos || eq == 0) continue;
                const std::string key = trim(line.substr(0, eq));
                if (std::getenv(key.c_str()) != nullptr) continue;  // real env vars win
                const std::string value = trim(line.substr(eq + 1));
#ifdef _WIN32
                ::_putenv_s(key.c_str(), value.c_str());
#else
                ::setenv(key.c_str(), value.c_str(), /*overwrite=*/0);
#endif
            }
            return;
        }
        if (dir == dir.root_path() || dir.parent_path() == dir) return;
    }
}

std::string env(const std::string& name) {
    const char* value = std::getenv(name.c_str());
    if (value == nullptr || *value == '\0') {
        throw std::runtime_error("Missing env var " + name +
                                 " - copy .env.example to .env and fill it in");
    }
    return value;
}

std::string replace_all(std::string s, const std::string& from, const std::string& to) {
    for (std::size_t pos = 0; (pos = s.find(from, pos)) != std::string::npos; pos += to.size()) {
        s.replace(pos, from.size(), to);
    }
    return s;
}

}  // namespace

int main(int argc, char** argv) {
    try {
        load_env();

        const SigningScheme scheme = parse_scheme(argc > 1 ? argv[1] : "hmac");
        const std::string prefix = scheme == SigningScheme::Hmac ? "HMAC" : "ED25519";

        HandoffOptions options;
        options.base_url = env("API_BASE_URL");
        options.scheme = scheme;
        options.secret = scheme == SigningScheme::Hmac
                             ? env("HMAC_SECRET")
                             // .env can only carry the PEM with real newlines escaped as literal "\n".
                             : replace_all(env("ED25519_PRIVATE_KEY_PEM"), "\\n", "\n");
        options.partner_id = env(prefix + "_PARTNER_SLUG");
        options.partner_user_id = env(prefix + "_TEST_PARTNER_USER_ID");
        options.phone_number = env(prefix + "_TEST_PHONE_NUMBER");

        const SignedResponse response = PartnerClient().handoff(options);

        std::cout << "status: " << response.status << "\n";
        if (response.ok) {
            if (auto result = response.data_as<HandoffResult>()) {
                std::cout << "entryUrl: " << result->entry_url << " (expires in "
                          << result->expires_in << "s)\n";
            }
        }
        std::cout << "body: " << (response.json ? response.json->dump(2) : response.raw_body) << "\n";
        return response.ok ? EXIT_SUCCESS : EXIT_FAILURE;
    } catch (const std::exception& error) {
        std::cerr << "Request failed: " << error.what() << "\n";
        return EXIT_FAILURE;
    }
}

File: rust/examples/handoff.rs
//! Rust equivalent of examples/test-handoff-hmac.js and examples/test-handoff-ed25519.js.
//! Reads the same repo-root .env as the Node examples.
//!
//!   cargo run --example handoff -- hmac
//!   cargo run --example handoff -- ed25519

use std::env;
use std::fs;
use std::process::ExitCode;

use partner_signed_client::{HandoffOptions, HandoffResult, PartnerClient, SigningScheme};

fn main() -> ExitCode {
    // Load .env before the async runtime starts any threads (set_var is only sound single-threaded).
    load_env();

    let result = tokio::runtime::Runtime::new()
        .map_err(Into::into)
        .and_then(|runtime| runtime.block_on(run()));

    match result {
        Ok(()) => ExitCode::SUCCESS,
        Err(error) => {
            eprintln!("Request failed: {error}");
            ExitCode::FAILURE
        }
    }
}

async fn run() -> Result<(), Box<dyn std::error::Error>> {
    let scheme: SigningScheme = env::args().nth(1).as_deref().unwrap_or("hmac").parse()?;
    let prefix = match scheme {
        SigningScheme::Hmac => "HMAC",
        SigningScheme::Ed25519 => "ED25519",
    };

    let secret = match scheme {
        SigningScheme::Hmac => var("HMAC_SECRET")?,
        // .env can only carry the PEM with real newlines escaped as literal "\n".
        SigningScheme::Ed25519 => var("ED25519_PRIVATE_KEY_PEM")?.replace("\\n", "\n"),
    };

    let options = HandoffOptions::new(
        var("API_BASE_URL")?,
        scheme,
        secret,
        var(&format!("{prefix}_PARTNER_SLUG"))?,
        var(&format!("{prefix}_TEST_PARTNER_USER_ID"))?,
        var(&format!("{prefix}_TEST_PHONE_NUMBER"))?,
    );

    let response = PartnerClient::new().handoff(options).await?;

    println!("status: {}", response.status);
    if response.ok {
        if let Ok(result) = response.data_as::<HandoffResult>() {
            println!(
                "entryUrl: {} (expires in {}s)",
                result.entry_url, result.expires_in
            );
        }
    }
    match &response.json {
        Some(json) => println!("body: {}", serde_json::to_string_pretty(json)?),
        None => println!("body: {}", response.raw_body),
    }
    Ok(())
}

fn var(name: &str) -> Result<String, String> {
    match env::var(name) {
        Ok(value) if !value.is_empty() => Ok(value),
        _ => Err(format!(
            "Missing env var {name} — copy .env.example to .env and fill it in"
        )),
    }
}

/// Minimal .env loader — walks up from the working directory to the first .env; real env vars win.
fn load_env() {
    let Ok(cwd) = env::current_dir() else { return };
    let Some(contents) = cwd
        .ancestors()
        .find_map(|dir| fs::read_to_string(dir.join(".env")).ok())
    else {
        return;
    };

    for line in contents.lines() {
        let line = line.trim();
        if line.starts_with('#') {
            continue;
        }
        let Some((key, value)) = line.split_once('=') else {
            continue;
        };
        let key = key.trim();
        if key.is_empty() || env::var_os(key).is_some() {
            continue;
        }
        // SAFETY: runs at the top of `main`, before any other thread exists.
        unsafe { env::set_var(key, value.trim()) };
    }
}
